### PR TITLE
Ideas about better db unittest test tooling

### PR DESCRIPTION
**Pull request by babky-atteq**
After little pair programming session with Martin we came with some
better DbTestCase, but it still needs more polish.

### CODE REVIEW

**Comment by babky-atteq** [comment]
Documentation for unit testing is added as well. It is now prepared for a review.

File: tests/resource/transactional_guard_test.py [hunk 1]
```diff
@@ -66,15 +67,17 @@ class DenyPolicy(ResourcePolicy):
...
-        resource.read('', {}, {})
+        with mock.patch('zsl.service.service.TransactionHolder',
+                        side_effect=TestTHolder):
+            resource = GuardedUserModel()
+            resource.read('', {}, {})
 
-        resource._orm.rollback.assert_called()
-        resource._orm.query.assert_not_called()
+        TestTHolder.rollback.assert_called()
+        TestTHolder._orm.assert_not_called()
```
Comment (morihladko) [inline tests/resource/transactional_guard_test.py]: Shouldn't this use `mock_db_session` too?

File: zsl/testing/http.py [hunk 2]
```diff
@@ -41,8 +46,8 @@ def requestTask(self, client, task, data, headers=None):
         :param task: Url which will be requested using POST method.
         :param data: Data which will be posted and first converted to JSON.
         :param headers: Dictionary of headers that'll be appended the
```
Comment (morihladko) [inline zsl/testing/http.py]: Should be there an '\' at the end of the line for param help string, when you wrap the line, so the sphinx includes it with it?

File: zsl/testing/zsl.py [hunk 3]
```diff
@@ -1,36 +1,51 @@
...
 from zsl.configuration import InvalidConfigurationException
 
 ZslTestConfiguration = namedtuple('ZslTestConfiguration',
                                   ['app_name', 'version', 'container',
-                                   'profile'])
+                                   'profile', 'config_object'])
+ZslTestConfiguration.__new__ = partial(ZslTestConfiguration.__new__,
+                                       config_object=None,
+                                       version=None,
+                                       profile=None)
```
Comment (morihladko) [inline zsl/testing/zsl.py]: :+1:

File: zsl/testing/db.py [hunk 4]
```diff
@@ -1,17 +1,101 @@
...
+        # type: (SessionFactory)->None
+        sess = session_factory.create_session()
+        sess.rollback()
+        sess.close()
 
+    @classmethod
+    @inject(session_factory=TestSessionFactory)
+    def tearDownClass(cls, session_factory):
+        session = session_factory.create_session()
+        session.rollback()
```
Comment (morihladko) [inline zsl/testing/db.py]: Is this here for ensuring the closing of the session, if some test method did not close it? Or is there another reason?

File: zsl/testing/zsl.py (hunk 3 above)
Comment (babky) [inline zsl/testing/zsl.py]: I'll write som docs on this.

File: tests/resource/transactional_guard_test.py (hunk 1 above)
Comment (babky-atteq) [inline tests/resource/transactional_guard_test.py]: No, it works for SessionFactory not for TxHolder

File: zsl/testing/http.py (hunk 2 above)
Comment (babky-atteq) [inline zsl/testing/http.py]: the result seems correct, am I missing something?

Comment (morihladko) [inline zsl/testing/http.py]: No, it's just my foggy memory about sphinx.

**Pull request merged**